src/metric-engine/src/row_modifier.rs

```diff
@@ -0,0 +1,493 @@
+            datatype: ColumnDataType::Uint64 as i32,
+            semantic_type: SemanticType::Tag as _,
+            datatype_extension: None,
+            options: None,
+        });
+        for iter in iter.iter_mut() {
+            let (table_id, tsid) = self.fill_internal_columns(table_id, &iter);
+            iter.row.values.push(table_id);
+            iter.row.values.push(tsid);
+        }
+
+        Ok(iter.rows)
+    }
+
+    /// Fills internal columns of a row with table name and a hash of tag values.
+    fn fill_internal_columns(&self, table_id: TableId, iter: &RowIter<'_>) -> (Value, Value) {
+        let mut hasher = mur3::Hasher128::with_seed(TSID_HASH_SEED);
+        for (name, value) in iter.primary_keys_with_name() {
+            // The type is checked before. So only null is ignored.
+            if let Some(ValueData::StringValue(string)) = &value.value_data {
+                name.hash(&mut hasher);
+                string.hash(&mut hasher);
+            }
+        }
+        // TSID is 64 bits, simply truncate the 128 bits hash
+        let (hash, _) = hasher.finish128();
+
+        (
+            ValueData::U32Value(table_id).into(),
+            ValueData::U64Value(hash).into(),
+        )
+    }
+}
+
+/// Index of a value.
+#[derive(Debug, Clone, Copy)]
+struct ValueIndex {
+    column_id: ColumnId,
+    index: usize,
+}
+
+/// Index of a row.
+struct IterIndex {
+    indices: Vec<ValueIndex>,
+    num_primary_key_column: usize,
+}
+
+impl IterIndex {
+    fn new(row_schema: &[ColumnSchema], name_to_column_id: &HashMap<String, ColumnId>) -> Self {
+        let mut reserved_indices = SmallVec::<[ValueIndex; 2]>::new();
+        let mut primary_key_indices = BTreeMap::new();
+        let mut field_indices = SmallVec::<[ValueIndex; 1]>::new();
+        let mut ts_index = None;
+        for (idx, col) in row_schema.iter().enumerate() {
+            match col.semantic_type() {
+                SemanticType::Tag => match col.column_name.as_str() {
+                    DATA_SCHEMA_TABLE_ID_COLUMN_NAME => {
+                        reserved_indices.push(ValueIndex {
+                            column_id: ReservedColumnId::table_id(),
+                            index: idx,
+                        });
+                    }
+                    DATA_SCHEMA_TSID_COLUMN_NAME => {
+                        reserved_indices.push(ValueIndex {
+                            column_id: ReservedColumnId::tsid(),
+                            index: idx,
+                        });
+                    }
+                    _ => {
+                        primary_key_indices.insert(
+                            col.column_name.as_str(),
+                            ValueIndex {
+                                column_id: *name_to_column_id.get(&col.column_name).unwrap(),
+                                index: idx,
+                            },
+                        );
+                    }
```

Comment:
Can we ensure the column order in `row_schema` has the same order of the primary key?